From 3c4f9dbf09c7283c6a796f709eb42e0b9eb6d7cf Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Sat, 16 May 2020 10:01:08 +0300
Subject: [PATCH] [SW-11875]: goya2: onix: add fuse read support

Mark fuse read function as weak and add implementation for onix
platform to return default value for SYSCFG word. This is required in
case verification of preboot images fails while using arcboot function

Change-Id: Ia744820a98e258ed2223d71ab66d03212b10463f
---

diff --git a/common/efuse.c b/common/efuse.c
index febb5b9..49951ca 100644
--- a/common/efuse.c
+++ b/common/efuse.c
@@ -48,7 +48,7 @@
 	return 0;
 }
 
-int preboot_fuse_read(u32 word, u32 *val)
+__attribute__((weak)) int preboot_fuse_read(u32 word, u32 *val)
 {
 	int offset, ret;
 	u32 read_val;
diff --git a/common/efuse.h b/common/efuse.h
index a967633..e4eedc4 100644
--- a/common/efuse.h
+++ b/common/efuse.h
@@ -12,6 +12,7 @@
 
 #define EFUSE_READ_RANGE_ALL		0x3F80
 #define EFUSE_PRODUCT_ID_OFFSET		0x1
+#define EFUSE_SYSCFG_OFFSET             0x3A
 
 int preboot_fuse_init(void);
 int preboot_fuse_read(u32 word, u32 *val);
diff --git a/goya2_arc/common.h b/goya2_arc/common.h
index 4913c1a..20157cb 100644
--- a/goya2_arc/common.h
+++ b/goya2_arc/common.h
@@ -18,6 +18,12 @@
 #define FUSE_WORDS_PER_BANK	0x40
 #define I2C_SLAVE_ADDRESS	0x6A
 
+#define EFUSE_SYSCFG_DISABLE_TWI_IF_MASK        BIT(19)
+#define EFUSE_SYSCFG_FW_VERSION_LOCATION_SPI	BIT(14)
+#define EFUSE_SYSCFG_KEY_LOCATION_SPI		BIT(8)
+#define EFUSE_SYSCFG_REVOKATION_LOCATION_SPI	BIT(6)
+#define EFUSE_SYSCFG_ENABLE_VALIDATION_MASK	BIT(10)
+
 /*
  * extracts "num" number of bits from position "pos" in the
  * given number "value"
diff --git a/goya2_arc/onix.c b/goya2_arc/onix.c
index 40a03da..110e7e4 100644
--- a/goya2_arc/onix.c
+++ b/goya2_arc/onix.c
@@ -7,12 +7,21 @@
 #include <preboot.h>
 #include <pll.h>
 #include <uart.h>
+#include <efuse.h>
 
 #include <psoc_global_conf_regs.h>
+
+#include "common.h"
 
 #define SYS_CLOCK_FREQUENCY	35000000ULL
 #define UART_CLOCK_FREQUENCY	SYS_CLOCK_FREQUENCY
 #define PLDM_UART_BAUDRATE	57600
+
+#define EFUSE_SYSCFG_ONIX_DEFAULT \
+				(EFUSE_SYSCFG_FW_VERSION_LOCATION_SPI | \
+				 EFUSE_SYSCFG_KEY_LOCATION_SPI | \
+				 EFUSE_SYSCFG_REVOKATION_LOCATION_SPI | \
+				 EFUSE_SYSCFG_ENABLE_VALIDATION_MASK)
 
 u32 hl_skip_pcie_init(void)
 {
@@ -39,4 +48,16 @@
 	return (factor + 1) & 0xFFFE;
 }
 
+int preboot_fuse_read(u32 word, u32 *val)
+{
+	int ret = -1;
+
+	if (word == EFUSE_SYSCFG_OFFSET) {
+		*val = EFUSE_SYSCFG_ONIX_DEFAULT;
+		ret = 0;
+	}
+
+	return ret;
+}
+
 void preboot_init_pcie_phy(void) {}
